### ISSUE TITLE
Bug: can't change and override valid locales

### ISSUE THREAD

**Issue by davislasis** [codeigniter4/CodeIgniter4#4297]
**Problem**
System does not change incoming request locale with overriding it from outer source (not App config, like database) using filter approach

**Describe the bug**
- all my routes has {locale} in front as first url segment 'en/dashboard', 'lv/dashboard' etc
- supported locales are stored in local mysql database and managed through admin area
- my project will populate / manage languages only through database and admin area, no need of App config $supportedLocales as client will not change and re-commit files
- to avoid caching locale change in frontend, i have setup mainFilter which caches potential locale change at any point (route) within the system
- IncomingRequest class is initialised before filters, so if my filter override supportedLocales - it does not change in further system loding

**CodeIgniter 4 version**
4.1.1

Solution 1:
- IncomingRequest class - variable `$validLocales` change from `protected` to `public`
- this way it is possible to override values if needed

**Affected module(s)**
/system/HTTP/IncomingRequest.php

[https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L99](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L99)

Solution 2:
- as App config does not have `$validLocales` variable
- as IncomingRequest class - variable `$validLocales` is not heavily used elsewhere
- as IncomingRequest class controller mandatory requests App config as first method param
- remove variable `$validLocales` from IncomingRequest class
- as IncomingRequest class method setLocale() uses `$validLocales`, replace it with already loaded `$this->config->supportedLocales`

**Affected module(s)**
/system/HTTP/IncomingRequest.php

remove: [https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L99](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L99)

remove: [https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L162](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L162)

update: [https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L225](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/IncomingRequest.php#L225)

```php

	/**
	 * Sets the locale string for this request.
	 *
	 * @param string $locale
	 *
	 * @return IncomingRequest
	 */
	public function setLocale(string $locale)
	{
		// If it's not a valid locale, set it
		// to the default locale for the site.
		if (! in_array($locale, $this->config->supportedLocales))
		{
			$locale = $this->defaultLocale;
		}

		$this->locale = $locale;

		// If the intl extension is loaded, make sure
		// that we set the locale for it... if not, though,
		// don't worry about it.
		// this should not block code coverage thru unit testing
		// @codeCoverageIgnoreStart
		try
		{
			if (class_exists('\Locale', false))
			{
				\Locale::setDefault($locale);
			}
		}
		catch (\Exception $e)
		{
		}
		// @codeCoverageIgnoreEnd

		return $this;
	}
```

**For reference my MainFilter class**
```php
<?php namespace Modules\Core\Base\Filters;

use CodeIgniter\HTTP\RequestInterface;
use CodeIgniter\HTTP\ResponseInterface;
use CodeIgniter\Filters\FilterInterface;
use Config\App;
use Config\Database;

class MainFilter implements FilterInterface
{
    /**
     * Do whatever processing this filter needs to do.
     * By default it should not return anything during
     * normal execution. However, when an abnormal state
     * is found, it should return an instance of
     * CodeIgniter\HTTP\Response. If it does, script
     * execution will end and that Response will be
     * sent back to the client, allowing for error pages,
     * redirects, etc.
     *
     * @param \CodeIgniter\HTTP\RequestInterface $request
     *
     * @param null $arguments
     * @return mixed
     */
	public function before(RequestInterface $request, $arguments = null)
	{
        $locales = [];

	    // get all languages from database
        $db = Database::connect();

        $results = $db->table('core_languages')
            ->select('language')
            ->where('published', 1)
            ->get()
            ->getResultArray();

        if (!empty($results)) {
            foreach ($results as $result) {
                $locales[] = $result['language'];
            }
        }

        if (!empty($locales)) {
            /** @var App $config */
            $config = Config('App');
            $config->supportedLocales = $locales;

            $request->config = $config;

            $locale = $request->uri->getSegment(1);

            if (in_array($locale, $locales)) {
                $request->setLocale($locale);
            } else {
                $request->setLocale($config->defaultLocale);
            }
        }
	}

	//--------------------------------------------------------------------

    /**
     * Allows After filters to inspect and modify the response
     * object as needed. This method does not allow any way
     * to stop execution of other after filters, short of
     * throwing an Exception or Error.
     *
     * @param \CodeIgniter\HTTP\RequestInterface $request
     * @param \CodeIgniter\HTTP\ResponseInterface $response
     *
     * @param null $arguments
     * @return mixed
     */
	public function after(RequestInterface $request, ResponseInterface $response, $arguments = null)
	{

	}

	//--------------------------------------------------------------------
}
```

**Context**
 - OS: Ubuntu 20.04 LTS
 - Web server: Apache 2.4.41
 - PHP version 7.4 and 8.0

**My suggestion**
I vote for solution 2 my-self and i can create a PR if needed after discussion and decision

**Comment by paulbalandan** [comment]
As `$validLocales` is protected, user extensions of `IncomingRequest` might be using this property in their own implementations which will break if we remove that. I'll vote partially for 2 with changing only the `setLocale` method by replacing the locales array in `in_array` and still set to strict.
```php
		// If it's not a valid locale, set it
		// to the default locale for the site.
		if (! in_array($locale, $this->config->supportedLocales, true))
		{
			$locale = $this->defaultLocale;
		}
```

**Comment by davislasis** [comment]
Great call on potential external user extensions to be envolved. I didn't think about that in first place.

I agree with using strict

I think `$validLocales` is not much used from outside and we could remove it and call it as "breaking change", although safer is not to do that.
From other hand, if CI has plans on this variable - then leave it as is for now, if not - then what's the point to not remove it?

Are we waiting on other opinions and/or comments?
or
Are we set and i can go with pull request?

**Comment by paulbalandan** [comment]
Please send in a PR. Maybe we'll get comments in the PR. For `$validLocales` we can deprecate it later if found it to be of no use.

**Comment by davislasis** [comment]
Great, thanks,
I will try to get some free time on weekend to get PR in.

**Comment by MGatner** [comment]
As I mentioned on the PR, I think your best bet will be to define a Registrar for `App` that supplies the locales from your database. This may require some testing since `App` is loaded pretty early and is a bit of a "special case", but it currently holds the definitive source of application locales so solutions that make `IncomingRequest` "exempt" from that seem unwise. Now I'm not opposed to thinking bigger about how we handle all these (maybe locales should be their own config??) but for now I think we're bound by backwards-compatibility.

**Comment by MGatner** [comment]
A third "quick solution" would be to define your own extension of `IncomingRequest` in `Config\Services`. That should be fairly straightforward:
```
	public static function request(App $config = null, bool $getShared = true)
	{
		if ($getShared)
		{
			return static::getSharedInstance('request', $config);
		}

		$config = $config ?? config('App');

		return new MyIncomingRequest(
			$config,
			static::uri(),
			'php://input',
			new UserAgent()
		);
	}

class MyIncomingRequest extends \CodeIgniter\HTTP\IncomingRequest
{
	/**
	 * Constructor
	 *
	 * @param object      $config
	 * @param URI         $uri
	 * @param string|null $body
	 * @param UserAgent   $userAgent
	 */
	public function __construct($config, URI $uri = null, $body = 'php://input', UserAgent $userAgent = null)
	{
		parent::__construct($config, $uri, $body, $userAgent);

		// Insert your locales database code here for $this->validLocales
	}
}
```

**Comment by MGatner** [comment]
@davislasis I have no problems whatsoever with the premise! I want it done right. If you want to submit a PR with Paul's suggestion so we can see what that actually looks like please do.

This problem is not unique: we have requests to load configuration from the database all the time. **Registrars** are really the tool for this, and they are much simpler than override core services (though I still think that is a viable and logical solution, but I haven't actually tried it).

**Comment by davislasis** [comment]
OK, example of PR change is upper in the thread and i already made a PR, only that your closed (deleted) it.
I can re-submit it again if it is not possible to revert it

Solution is 
https://github.com/codeigniter4/CodeIgniter4/blob/b3ebd3da92b9da3971fb02d912f3923939b9c008/system/HTTP/IncomingRequest.php#L224

change it to
`if (! in_array($locale, $this->config->supportedLocales, true))`


Thanks,
Dāvis

**Comment by MGatner** [comment]
Oh right, #4478. We can always reopen. I guess I misunderstood what @paulbalandan was saying, I still don't think that is the right solution because it would invalidate anything done in `$this->validLocales`. Maybe a compromise for now would be to check both?
```
$locales = array_merge($this->validLocales, $this->config->supportedLocales);
 if (! in_array($locale, $locales, true))
...
```

Let's see what @paulbalandan says because I didn't know he had addressed this one before I saw your PR.

**Comment by davislasis** [comment]
OK, let's wait on @paulbalandan 's decision

Meanwhile, your merge stuff works, but maybe fully avoid possible conflicts with original core flow, we could implement 2nd param as toggle? Something along those lines:

public function setLocale(string $locale, bool $override = false)
{
    $locales = $this->validLocales;

    if ($override)
    {
        $locales = array_merge($this->validLocales, $this->config->supportedLocales);
    }

    if (! in_array($locale, $locales, true))
    {
        $locale = $this->defaultLocale;
    }

    $this->locale = $locale;
    Locale::setDefault($locale);

    return $this;
}

**Comment by paulbalandan** [comment]
I haven't thought that while `$validLocales` is unused anywhere except for the constructor and setLocale, it may be used in userland. So, yeah, removing that fully in the picture might invalidate the stuff made to it. Using registrars seems to me to be complicated (or maybe I just don't know where to hook the registrar call and haven't done it yet).

It seems to me in order to resolve both sides we'd better go with the `array_merge` solution for now until we get to a deeper fix for this, and a verdict on `$validLocales` array. My concern is we cannot use `$this->config` as the overriding config. This must be freshly supplied. The fix only works because you have edited directly the shared instance of `App` in your filter. If someone will be using a new instance, then `$this->config` will still be pointing to the old config.

My suggestion is something like this:
```php
public function setLocale(string $locale, ?App $override = null)
{
    if (! is_null($override))
    {
        $this->validLocales = array_merge($this->validLocales, $override->supportedLocales);
    }

    if (! in_array($locale, $this->validLocales, true))
    {
        $locale = $this->defaultLocale;
    }

    ...
}
```

**Comment by davislasis** [comment]
Thanks @paulbalandan for quick response. 
Brainstorming further, i now understand the possible issue with "old" config stuff and the best i guess would be "the fresh data".
So, i think 2nd param should simply be array, not entire App Config class where we use only `supportedLocales`.

`
public function setLocale(string $locale, array $locales = [])
{
    if (! empty($locales))
    {
        $this->validLocales = array_merge($this->validLocales, $locales);
    }

    if (! in_array($locale, $this->validLocales, true))
    {
        $locale = $this->defaultLocale;
    }

    ...
}
`

I would like to hear a last verdict on how we change this: this option or @paulbalandan 's ?

**Comment by paulbalandan** [comment]
Supplying an array seems faster and more economical to me.

**Comment by MGatner** [comment]
I like this solution, with a second array of supported locales. Do also look into Registrars, they are a pretty awesome component. Right now your workflow looks like this:
1. App loads Config
2. Route launches filter
3. Filter loads locales
4. Filter adds locales to Config
5. Filter replaces Request locales with Config
6. Filter sets Request locale

With a Registrar this becomes:
1. App loads Config, triggers Registrar
2. Registrar loads locales, adds to Config
3. Filter sets Request locale

Registrars are very easy - do read the [User Guide](https://codeigniter4.github.io/CodeIgniter4/general/configuration.html#registrars ) but I think it makes them seem more complicated. In your case you would create the file **app/Config/Registrars.php** and add a bee method named for the Config file you are targeting that returns an array of the data to merge. For example:
```
class Registrar
{
public function app()
{
// load locales from database 
return ['supportedLocales' => $localesFromDatabase];
}
}
```

Please excuse formatting and possible errors as I am mobile.

**Comment by fana605** [comment]
> n your case you would create the file **app/Config/Registrars.php** and add a static method named for the Config file you are targeting that

I needed this exact same functionality and after reading @MGatner 's tip about Registrars and going through the documentation, I can confirm this works perfectly! I was not aware that these Registrars existed (the docs are very discrete about it), but this really opens a lot of possibilities around overriding config options from the database.
Thank you, @MGatner, and in my opinion, there's no need to change setLocale() method in IncommingRequest Class  (#4513).

**Comment by MGatner** [comment]
Thanks for the feedback @fana605! I am glad to hear it worked - App is a bit of a special case since it is needed before the others so I wasn't 100% sure it would work as anticipated. Registrars are very powerful and definitely deserve better User Guide coverage. 

My stance hasn't changed, I still think Registrars is the appropriate solution for this issue but I'm willing to accept the PR as a compromise.

**Comment by fana605** [comment]
> Thanks for the feedback @fana605! I am glad to hear it worked - App is a bit of a special case since it is needed before the others so I wasn't 100% sure it would work as anticipated. Registrars are very powerful and definitely deserve better User Guide coverage.
> 
> My stance hasn't changed, I still think Registrars is the appropriate solution for this issue but I'm willing to accept the PR as a compromise.

Just as a follow up, I can access the Database on the Registrar but not the Session. Not sure what the problem is, maybe something on my side, but we need to make sure all the functionality is available at the time the Registrar runs.

**Comment by MGatner** [comment]
@fana605 that doesn't surprise me. I would have to dig a bit to determine the reason, but `Config\App` is a bit of a special case like I mentioned. In most cases **bootstrap.php** will call the CodeIgniter service: https://github.com/codeigniter4/CodeIgniter4/blob/b19061649f632aa74a8293d75c1b195357560247/system/bootstrap.php#L156
... which makes the first call to `Config\App`: https://github.com/codeigniter4/CodeIgniter4/blob/f19fe3627aeaa46fc77c55305e490ab49e46c02d/system/Config/Services.php#L152

This all takes place before `CodeIgniter` is initialized and bootstrapped, so some services might not be "readied" like we are used to. That said, CI4 is compartmentalized pretty well so I expect there is a way to get Sessions working in the `App` registrar without much trouble.

**Comment by fana605** [comment]
@MGatner , if you have any tips about getting the session to work on the registrar, I'm grateful. It seems that even the superglobal $_SESSION is not available at that point.

**Comment by MGatner** [comment]
Could you share your code that didn't work? I will try to recreate this on Playground and see if I can track down what needs to happen.

**Comment by fana605** [comment]
> Could you share your code that didn't work? I will try to recreate this on Playground and see if I can track down what needs to happen.

Something as simple as this will give me a fatal error "allowed memory exhausted". Maybe it generates some infinite loop...

`class Registrars
{
    public static function App()
    {
        $session = \Config\Services::session();
    }
}`

**Comment by MGatner** [comment]
Ah yes that's precisely it: Session relies on `Config\App` so trying to initialize it during the `App` registrar creates an endless loop. Probably if all you need is to grab a session variable you would be fine to handle the session yourself:
```
session_start();
$myVar = $_SESSION['myVar'] ?? 'FallbackValue';
```

**Comment by davislasis** [comment]
@fana605 can you share your registrar approach code that did work, please? 

My test somewhat worked, url and route, no infinite loops, but landing in a new locale page, still does not load "requested" locale, but default EN.
I did not have time to dig through on holidays, but I guess this could be something simillar as for sessions, wher App config is re-loaded through registrar and therefor load a new "default" language instance and language files as service, meaning shared and no infinite loop, so we actually don't see a problem's source.

hope this makes some sence

**Comment by fana605** [comment]
> @fana605 can you share your registrar approach code that did work, please?
> 
> My test somewhat worked, url and route, no infinite loops, but landing in a new locale page, still does not load "requested" locale, but default EN.
> I did not have time to dig through on holidays, but I guess this could be something simillar as for sessions, wher App config is re-loaded through registrar and therefor load a new "default" language instance and language files as service, meaning shared and no infinite loop, so we actually don't see a problem's source.
> 
> hope this makes some sence

Hi, @davislasis , sure.

On my Config/App.php (at the end):

```
public static $registrars = [
	Registrars::class
];
```

Then, my Config/Registrars.php:

```
namespace Config;

class Registrars
{
    public static function App()
    {
        $db = \Config\Database::connect();

        $builder = $db->table('locales')
            ->select('locales.*')
            ->where('publicado', 1)
            ->orderBy('ordem', 'asc');

        $locales = $builder->get()->getResult();

        foreach ($locales as $locale) {
            $codes = array_map('trim', explode(',', $locale->codes));
            $langs[$locale->abbrev] = [
                'id' => $locale->id,
                'abbrev' => $locale->abbrev,
                'name' => $locale->lang,
                'folder' => $locale->folder,
                'default' => (int)$locale->default,
                'codes' => array_filter($codes)
            ];
        }

        $defaultLocale = '';
        foreach ($langs as $locale) {
            if ($locale['default'] === 1) {
                $defaultLocale = $locale['abbrev'];
                break;
            }
        }

        return [
            'defaultLocale' => $defaultLocale,
            'supportedLocales' => array_keys($langs)
        ];
    }
}
```

But this was not enough, actually. I then have a Filter that I use to set the locales, and I had to add this:

```
namespace App\Filters;

use CodeIgniter\HTTP\RequestInterface;
use CodeIgniter\HTTP\ResponseInterface;
use CodeIgniter\Filters\FilterInterface;

class NegociateLocale implements FilterInterface
{
    public function before(RequestInterface $request, $arguments = null)
    {
        //Other code...

        //I had to add this:
        $request->setLocale(config('App')->defaultLocale);
    }

    public function after(RequestInterface $request, ResponseInterface $response, $arguments = null)
    {
    }
}
```

This makes me doubt that the Registrar is really a good solution for this case. Even more because I can't use sessions and I'm using database sessions... This doesn't feel like a natural, future proof solution to me...
I guess I'll try a different approach.
What do you think @davislasis ?

**Comment by davislasis** [comment]
Hi @fana605 , thanks for your feedback.

I agree that it does not feel natural.
I don't like the idea that you need to load whole App config "again".

As @MGatner said - let's use this as a compromise for now. If anyone need to override his own locales, it can be done through filter (my example of Mainfilter on top of this issue article), forcing his locales as 2nd param of IncomingRequest->setLocale()

Let's use this for now. Meanwhile anyone (everyone) of us should dig through this and figure out (fix, modify etc) Registrar (or other) approach to simplify this kinda simple issue. 
Maybe think of a more global approach so that could work on any other config value to be overrided? 
Maybe change behaviour of {locale} in routes and in route collection class?

**Comment by fana605** [comment]
Hi @MGatner, is this fix going to be committed soon? I mean, the Registrar way is not working in my case because I can't access session. So, the suggested fix by @davislasis works perfectly for me and I am now using a modified version of system/HTTP/IncomingRequest.php, but that's obviously not the best option.
Do you guys think this will get a solution soon?

**Comment by MGatner** [comment]
Of you are referring to #4513 I think that the content there was fine, not sure why it never merged. At this point the PR would need to be rebased or recreated and sent to `4.2` branch because `develop` is locked to style-only changes.

**Comment by kenjis** [comment]
How about this?

```php
--- a/system/HTTP/IncomingRequest.php
+++ b/system/HTTP/IncomingRequest.php
@@ -162,7 +162,7 @@ class IncomingRequest extends Request
         $this->uri          = $uri;
         $this->body         = ! empty($body) ? $body : null;
         $this->userAgent    = $userAgent;
-        $this->validLocales = $config->supportedLocales;
+        $this->validLocales =& $config->supportedLocales;
 
         parent::__construct($config);
 
```
```php
        /** @var \Config\App $config */
        $config = config('App');
        $config->supportedLocales = ['en', 'id'];

        $this->request->setLocale('id');
        dd($this->request->getLocale()); //  "id"
```

**Comment by davislasis** [comment]
Hello,
I am still keen on finishing and pushing through this.
At the latest after 20th of december i will have a holiday vacation. I will free up some time to install 4.2 and test out those suggestions.

p.s. PR merge was failing on DB tests, not on my piece of code, so it was rejected and abondoned for some time.
PHPUnit / PHP 7.3 - SQLSRV (pull_request) Failing after 1m — PHP 7.3 - SQLSRV

**Comment by davislasis** [comment]
Hello guys,

Long story short - Registrar's method DOES WORK !!!


I dig through Registrar functionality and found out that it's strange, but actually yet powerfull stuff.

9 month later - sorry @MGatner for lacking knowledge back in a days, just had no time to properly dig through...


Long story long
1. you don't need to change `system\HTTP\IncomingRequest.php`
2. you don't need to add $registrars variable in `Config\App.php` (@fana605)
3. you need to setup additional stuff. There is my setup:

3.1. ` App/Config/App.php public $defaultLocale = 'en'; // default repo value`
3.2. `App/Config/App.php public $supportedLocales = ['en']; // default repo value`
3.3. DB for locales (languages)
![image](https://user-images.githubusercontent.com/42682325/145395124-d7560890-14a1-4e0b-9e43-edbaf7e1bae0.png)

3.4. App\Config\Registrar.php

```php
namespace App\Config;

use Config\Database;

class Registrar
{
    public static function App(): array
    {
        $output = [];

        //$config = config('App'); // can't use - makes infinite loop obviously
        //$output['supportedLocales'] = $config->supportedLocales;

        // connect to database
        $db = Database::connect();

		// get all languages from database
        $results = $db->table('core_languages')
            ->select('language, default')
            ->where('published', 1)
            ->get()
            ->getResultArray();

        if (!empty($results)) {
            foreach ($results as $result) {
                $output['supportedLocales'][] = $result['language']; // Config\App->supportedLocales = ['en'];
                //$output['supportedLocales'][$result['language']] = $result['language']; // Config\App->supportedLocales = ['en' => 'en'];

                if (!empty($result['default'])) {
                	$output['defaultLocale'] = $result['language'];
                } 
            }
        }

        return $output;
    }
}
```

3.5. App\Filters\MainFilter.php

```php
namespace App\Filters;

use CodeIgniter\HTTP\RequestInterface;
use CodeIgniter\HTTP\ResponseInterface;
use CodeIgniter\Filters\FilterInterface;

class MainFilter implements FilterInterface
{
	public function before(RequestInterface $request, $arguments = null)
	{
        $config = config('App');
        $config->supportedLocales = array_flip(array_flip($config->supportedLocales));

        $request->setLocale($config->defaultLocale);
	}

	public function after(RequestInterface $request, ResponseInterface $response, $arguments = null)
	{

	}
}
```

This is built and tested in CodeIgniter 4.1.1.

There is an issue (the same issue will be in CI 4.2) - BaseConfig merges config values with Registrar values if that config variable is an array.
That makes situation: as i have `en` in App config and `en` comes from DB as well, final config's `$supportedLocales` has 2 `en` languages in array.

```php
$this->app->supportedLocales array (7)
⇄0 => string (2) "en"
⇄1 => string (2) "en"
⇄2 => string (2) "lv"
⇄3 => string (2) "lt"
⇄4 => string (2) "et"
⇄5 => string (2) "ru"
⇄6 => string (2) "xy"
```

There are 2 possible solutions / workarounds for this:

1. change `App/Config/App.php` to key-based array `public $supportedLocales = ['en' => 'en'];` This would count as `breaking change` and i don't encourage this.
2. use MainFilter line 12 to merge values around: `$config->supportedLocales = array_flip(array_flip($config->supportedLocales));`
3. possible other solutions???

Final

As promised, in upcoming weeks, i will install CI 4.2 and i will test this all out. After that i could create new PR for userguide repo [https://github.com/codeigniter4/userguide/tree/master/docs](https://github.com/codeigniter4/userguide/tree/master/docs)

I think those code examples and explanations could fit in Implicit Registrars section?
[https://codeigniter.com/user_guide/general/configuration.html#implicit-registrars](https://codeigniter.com/user_guide/general/configuration.html#implicit-registrars)

In general this issue could be closed, unless you have anything to add regards this.

Cheers,
Dāvis

**Comment by kenjis** [comment]
@davislasis 

> As promised, in upcoming weeks, i will install CI 4.2 and i will test this all out. After that i could create new PR for userguide repo https://github.com/codeigniter4/userguide/tree/master/docs

What do you mean CI 4.2? 4.2 does not exist yet. Try the latest `develop` branch.
And the userguide repo is read only. Please send a PR for user guide to this repo.

**Comment by MGatner** [comment]
That's great @davislasis ! I'm glad the Registrars are working - I love that component, very powerful and definitely underused. I've tried to improve the docs and recommend them whenever appropriate. 

You should also check out the new Settings library (https://github.com/codeigniter4/settings) which is designed to do precisely what you're after: load overriding Config values from persistent storage (like Database). This could circumvent the need for filters altogether.

**Comment by fana605** [comment]
Hi @davislasis, this still won't work for my case because I need to access Session in the Registrar and it's not available.
The solution @kenjis points is the one that works best for me:
`$this->validLocales =& $config->supportedLocales;`
I didn't change the `system\HTTP\IncomingRequest.php` file, I overrode the IncomingRequest construct using the `app/Config/System.php` file and it works perfectly. Of course, if the construct changes in the future, it may break this solution.

**Comment by kenjis** [comment]
It seems almost impossible with the current implementation to change the validLocales of IncomingRequests using Session in the `Config\App`. See https://github.com/codeigniter4/CodeIgniter4/issues/7308#issuecomment-1445520397

**Comment by kenjis** [comment]
I sent another PR #7309

**Comment by kenjis** [comment]
Closed by #7309
